cloudsync/youtube.py

```diff
@@ -0,0 +1,247 @@
+            media_body(MediaFileUpload): The file containing the captions, in VTT format
+            video_id(str): The YouTube ID of the video to associate the captions with.
+
+        Returns:
+            Request response in JSON format
+
+        """
+        request = self.client.captions().insert(
+            part="snippet",
+            body=dict(
+                snippet=dict(
+                    videoId=video_id,
+                    language=caption.language,
+                    name=caption.language_name,
+                    isDraft=False
+                )
+            ),
+            media_body=media_body
+        )
+        return resumable_upload(request)
+
+    def update_caption(self, media_body, caption_id):
+        """
+        Update an existing YouTube caption with a new file and return the JSON response.
+
+        Args:
+            media_body(MediaFileUpload): the video caption file to upload
+            caption_id(str): The YouTube ID of the caption
+
+        Returns:
+            the result of the request in JSON format.
+        """
+        request = self.client.captions().update(
+            part="snippet",
+            body=dict(
+                id=caption_id,
+                snippet=dict(
+                    isDraft=False
+                )
+            ),
+            media_body=media_body
+        )
+        return resumable_upload(request)
+
+    def delete_caption(self, caption_id):
+        """
+        Delete a caption from Youtube
+
+        Args:
+            caption_id(str): The ID of the YouTube caption
+
+        Returns:
+            204 status code if successful
+        """
+        return self.client.captions().delete(id=caption_id).execute()
+
+    def upload_video(self, video, privacy='unlisted'):
+        """
+        Transfer the video's original video file from S3 to YouTube.
+        The YT account must be validated for videos > 15 minutes long:
+        https://www.youtube.com/verify
+
+        Args:
+            video(Video): The Video object whose original source file will be uploaded'
+            privacy(str): The privacy level to set the YouTube video to.
+
+        Returns:
+            Request response in JSON format
+
+        """
+        videofile = video.videofile_set.filter(encoding=EncodingNames.ORIGINAL).first()
+
+        request_body = dict(
+            snippet=dict(
+                title=video.title,
+                description=video.description
+            ),
+            status=dict(
+                privacyStatus=privacy
+            )
+        )
+
+        # YouTube API seems to insist on the video contents being available in their entirety before uploading,
+        # so download from S3 into a temporary named file.
+        with NamedTemporaryFile() as uploadfile:
+            self.s3.download_file(settings.VIDEO_S3_BUCKET, videofile.s3_object_key, uploadfile.name)
+            request = self.client.videos().insert(
```

Comment:
If the insert fails is an exception raised here or is there something we need to do with the `request`?

Reply:
The insert doesn't actually upload anything to Youtube here (that happens with `resumable_upload`) - it is just creating a `googleapiclient.http.HttpRequest` that will be used as the input parameter for `resumable_upload`.  But if for some reason the creation of this object or the download from S3 fails, then an exception will be raised here and should show up on Sentry.